File: Algorithm Program/22-23/fall/lecture/math/2-prime-factorization/sum_of_divisors.py
# https://ideone.com/hsETTx
n = 10512969114312


def calc_geometric_sum(prime: int, exponent: int) -> int:
    return (prime ** (exponent + 1) - 1) // (prime - 1)


def calc_divisors_sum_without_formula(n: int) -> int:
    sum = 0
    for i in range(1, n + 1):
        # i * i is faster than i**2. I first realized while testing this code!
        if i * i > n:
            break
        if n % i:
            continue
        sum += i
        if i * i != n:
            sum += n // i
    return sum


def calc_divisors_sum_with_formula(n: int) -> int:
    sum = 1
    for i in range(2, n + 1):
        if i * i > n:
            break
        if n % i:
            continue
        exponent = 0
        while n % i == 0:
            n /= i
            exponent += 1
        sum *= calc_geometric_sum(i, exponent)
    if n > 1:
        sum *= calc_geometric_sum(n, 1)
    return sum


print("n =", n)
print("Sum of divisors w/ formula\t=", calc_divisors_sum_with_formula(n))
print("Sum of divisors w/o formula\t=", calc_divisors_sum_without_formula(n))

File: Algorithm Program/22-23/fall/lecture/math/2-prime-factorization/sum_of_divisors.cpp
// https://ideone.com/D2aEfj
#include <iostream>
using namespace std;
using ll = long long;

ll power(ll prime, int exponent) {
    ll res = 1;
    while (exponent--)
        res *= prime;
    return res;
}

// 1 + prime + prime^2 + ... + prime^exponent
ll calc_geometric_sum(ll prime, int exponent) {
    return (power(prime, exponent + 1) - 1) / (prime - 1);
}

ll calc_divisors_sum_without_formula(ll n) {
    ll sum = 0;
    for (int i = 1; 1ll * i * i <= n; i++) {
        if (n % i)
            continue;
        sum += i;
        if (i != n / i)
            sum += n / i;
    }
    return sum;
}

int main() {
    ll n = 10512969114312ll; // Put ll or LL to the end to specify 64-bit literal
    ll initial_n = n;
    ll sum_of_divisors_with_formula = 1;
    ll sum_of_divisors_without_formula = calc_divisors_sum_without_formula(n);
    for (int i = 2; 1ll * i * i <= n; i++) {
        if (n % i)
            continue;
        // i is a prime divisor
        int exponent = 0;
        while (n % i == 0) {
            n /= i;
            exponent++;
        }
        sum_of_divisors_with_formula *= calc_geometric_sum(i, exponent);
    }
    // n became a prime itself
    if (n > 1) {
        // Factorization includes n^1
        sum_of_divisors_with_formula *= calc_geometric_sum(n, 1);
    }
    cout << "n = " << initial_n << "\n";
    cout << "Sum of divisors w/ formula\t= " << sum_of_divisors_with_formula << "\n";
    cout << "Sum of divisors w/o formula\t= " << sum_of_divisors_without_formula << "\n";
}